### PR TITLE
Various SDK tweaks and fixes

### LINKED ISSUE #2115: Explore solutions for improving FallbackProvider reliability
Problems:
            - ~~Fallback provider aggregates the underlying errors which makes them hard to parse and produce large logs~~
            - Intermittent problems with the network connection can lead to failures
            - https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/031f0a9692ec1c631c8921fb25864aff4377f2c0/typescript/sdk/src/deploy/HyperlaneDeployer.ts#L91 shows null at times

Possible solutions:
            - ~~We should modify fallback provider to catch and identify the actual underlying error for better legibility~~
            - We should consider wrapping the fallback provider with retry providers as well to be more resilient to intermittent network

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Rename RpcPagination fields in ChainMetadata for clarity (useful for SmartProvider)
- Add optional maxBlockAge field to RpcPagination
- Use RetryProvider and StaticJsonRpcProvider by default in MultiProvider (See #2024)
- Improves RetryProvider and Fixes #2115
- Enable users to customize MultiProvider provider creation behavior
- Replace `ExplorerFamily` with template literal type for better compat with direct string value uses in ChainMetadata configs
- Improve ChainMetadata validation
- Add back hyperlaneContractAddresses removed in #2042

### Backward compatibility

No, small breaking changes:
- Rename  RpcPagination to RpcPaginationOptions
- Rename providerBuilder to defaultProviderBuilder
- Remove RetryProvider
- Remove unused events functions, including getEvents, queryAnnotatedEvents, Annotated, and TSContract

### CODE REVIEW

File: typescript/sdk/src/events.ts [hunk 1]
```diff
@@ -168,14 +168,14 @@ async function getPaginatedEvents<T extends TypedEvent>(
 ): Promise<Array<T>> {
```
Comment (jmrossy) [inline typescript/sdk/src/events.ts:168]: Note, I plan to delete this file entirely when we move SmartProvider to the SDK. Though I'm open to just killing it now if people prefer.

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ export enum ExplorerFamily {
   Blockscout = 'blockscout',
   Other = 'other',
 }
+export type ExplorerFamilyType = `${ExplorerFamily}`;
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:19]: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

This lets someone use the literal string values (e.g. `'etherscan'`) **or** the enum reference (`ExplorerFamily.Etherscan`) so people don't have to reference the enum type in configs.

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 3]
```diff
@@ -47,7 +48,7 @@ export interface ChainMetadata {
     url: string;
     apiUrl?: string;
     apiKey?: string;
-    family?: ExplorerFamily;
+    family?: ExplorerFamilyType;
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:53]: See above

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 4]
```diff
@@ -373,13 +376,18 @@ export const fuji: ChainMetadata = {
   name: Chains.fuji,
   displayName: 'Fuji',
   nativeToken: avaxToken,
-  publicRpcUrls: [{ http: 'https://api.avax-test.network/ext/bc/C/rpc' }],
+  publicRpcUrls: [
+    {
+      http: 'https://api.avax-test.network/ext/bc/C/rpc',
+      pagination: { maxBlockRange: 2048 },
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:391]: Because the RPC doesn't require one. minBlockNumber would mean it doesn't allow querying below a certain number

File: typescript/sdk/src/events.ts [hunk 5]
```diff
@@ -168,14 +168,14 @@ async function getPaginatedEvents<T extends TypedEvent>(
...
   const metadata = multiprovider.getChainMetadata(chain);
   const pagination = metadata.publicRpcUrls[0].pagination;
-  if (!pagination) {
-    throw new Error('Domain need not be paginated');
-  }
+  if (!pagination) throw new Error('Domain need not be paginated');
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/events.ts:171]: No, if an rpc config does not specify any pagination constraints, then it need not be paginated right?

File: typescript/sdk/src/events.ts [hunk 6]
```diff
@@ -168,14 +168,14 @@ async function getPaginatedEvents<T extends TypedEvent>(
...
+  if (!pagination.maxBlockRange) throw new Error('MaxBlockRange required');
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/events.ts:172]: It's required for these helper functions here (which as you said are unused vestiges) but not required for SmartProvider, which is ultimately the only thing that will use these configs

**Review by asaj — APPROVED** [review]

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 7]
```diff
@@ -38,16 +41,16 @@ export interface ChainMetadata {
   publicRpcUrls: Array<{
     http: string;
     webSocket?: string;
-    pagination?: RpcPagination;
-    retry?: RetryOptions;
+    pagination?: RpcPaginationOptions;
+    retry?: RetryProviderOptions;
```
Comment (nambrot) [inline typescript/sdk/src/consts/chainMetadata.ts:45]: I like that this is overridable, but IMO we should not expect developers to know how to override this, so I would personally suggest a default that does retry, maybe like once or twice with a second in between?

Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:45]: Yep that's how it'll work with MultiProvider after this
